src/Compilers/Core/Portable/StrongName/PortableStrongNameProvider.cs

```diff
@@ -0,0 +1,80 @@
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.Cci;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class PortableStrongNameProvider : StrongNameProvider
+    {
+        private readonly ImmutableArray<string> _keyFileSearchPaths;
+        internal StrongNameFileSystem FileSystem { get; private set; }
+
+        public PortableStrongNameProvider(ImmutableArray<string> keySearchPaths = default, StrongNameFileSystem strongNameFileSystem = default)
+        {
+            FileSystem = strongNameFileSystem ?? new StrongNameFileSystem();
+            _keyFileSearchPaths = keySearchPaths.NullToEmpty();
+        }
+
+        public override int GetHashCode()
+        {
+            return 0;
+        }
+
+        internal override SigningCapability Capability => SigningCapability.SignsPeBuilder;
+
+        internal override StrongNameKeys CreateKeys(string keyFilePath, string keyContainerName, CommonMessageProvider messageProvider)
+        {
+            var keyPair = default(ImmutableArray<byte>);
+            var publicKey = default(ImmutableArray<byte>);
+            string container = null;
+
+            if (!string.IsNullOrEmpty(keyFilePath))
+            {
+                try
+                {
+                    string resolvedKeyFile = FileSystem.ResolveStrongNameKeyFile(keyFilePath, _keyFileSearchPaths);
+                    if (resolvedKeyFile == null)
+                    {
+                        throw new FileNotFoundException(CodeAnalysisResources.FileNotFound, keyFilePath);
+                    }
+
+                    Debug.Assert(PathUtilities.IsAbsolute(resolvedKeyFile));
+                    var fileContent = ImmutableArray.Create(FileSystem.ReadAllBytes(resolvedKeyFile));
+                    return StrongNameKeys.CreateHelper(fileContent, keyFilePath);
+                }
+                catch (Exception ex)
+                {
+                    return new StrongNameKeys(StrongNameKeys.GetKeyFileError(messageProvider, keyFilePath, ex.Message));
+                }
+                // it turns out that we don't need IClrStrongName to retrieve a key file,
+                // so there's no need for a catch of ClrStrongNameMissingException in this case
+            }
+
+            return new StrongNameKeys(keyPair, publicKey, null, container, keyFilePath);
+        }
+
+        internal override void SignPeBuilder(ExtendedPEBuilder peBuilder, BlobBuilder peBlob, RSAParameters privkey)
+        {
+            peBuilder.Sign(peBlob, content => SigningUtilities.CalculateRsaSignature(content, privkey));
+        }
+
+        internal override Stream CreateInputStream()
+        {
+            throw new NotSupportedException();
+        }
+
+        public override bool Equals(object obj)
+        {
+            return Object.Equals(this, obj);
+        }
```

Comment:
`Equals` and `GetHashCode` can be removed.

Reply:
This changes the public API.

Reply:
Isn't `PortableStrongNameProvider` a new type?

Reply:
```
Error	CS0534	'PortableStrongNameProvider' does not implement inherited abstract member 'StrongNameProvider.Equals(object)'	CodeAnalysis	C:\workspace\roslyn\src\Compilers\Core\Portable\StrongName\PortableStrongNameProvider.cs	17	Active
```

Problem is that StrongNameProvider has this

```
        public abstract override int GetHashCode();
        public override abstract bool Equals(object other);
```